docs/hugo/content/design/ADR-2023-09-Skipping-Properties/_index.md

```diff
@@ -0,0 +1,418 @@
+* Changes the object model of released resources, a breaking change for code consumers of the object model.
+* When there are multi-generational skips, other questions arise: Do we introduce the new type into every intermediate version, only the last, or both first and last? Each approach has advantages and drawbacks.
+
+### Option 2: Back conversion by new version
+
+Ensure that complex types stored in property bags have a consistent shape by introducing a dedicated type for that purpose. Regardless of whether the property bag is populated by an earlier version (e.g. a **v3** `Address`) or a later version (e.g. a **v5** `Address`), the shape of the object stored property bag will be the same (**v4**).
+
+<!-- yuml.me
+
+[v3.Person|FullName string;FamilyName string; KnownAs string; ResidentialAddress v3.Address|Properties PropertyBag]
+[v3.Person]->[v3.Address|Label multilineString|Properties PropertyBag]
+
+[v4.Person|FullName string;FamilyName string; KnownAs string|Properties PropertyBag]
+[v4.Person]-.-[v4.Address {bg:cornsilk}|Label multilineString|Properties PropertyBag]
+
+[v5.Person|FullName string;FamilyName string; KnownAs string;ResidentialAddress v5.Address|Properties PropertyBag]
+[v5.Person]->[v5.Address|Street string; Suburb string; City string; Country string|Properties PropertyBag]
+
+-->
+
+{{< figure src="./option-2-class.png" >}}
+
+We create the shape of the intermediate type by replicating the structure of the version _before_ (**v3**), essentially just copying it forward to the new version (**v4**).
+
+When we convert from **v3** to **v4**, we first convert the **v3** `Address` into a **v4** `Address` before storing it in the property bag.
+
+Then, when converting from **v4** to **v5**, we rehydrate a **v4** `Address` from the property bag and then convert that to a **v5** `Address`.
+
+The process for conversion from **v5** to **v4** to **v3** is the reverse. We convert the **v5** `Address` into a **v4** `Address` before storing it in the property bag, and then rehydrate a **v4** `Address` from the property bag and convert that to a **v3** `Address`.
+
+While this might seem to just _kick the can down the road_, our conversion framework already has support for customization of the conversions between adjacent versions of a type, allowing the mismatch between **v4** and **v5** of `Address` to be handled by augmenting the conversion with a hand crafted conversion.
+
+Pros:
+* Addresses the central inconsistency problem, ensuring we always serialize the same shape into intermediate property bags
+* Preserves the serialization format for existing resource versions (e.g. **v3** and **v4**)
+
+Cons:
+* Changes the object model of released resources, a breaking change for consumers of the object model
```

Comment:
Does it actually do this? It seems to just add _types_ to existing packages, which is OK?

Those types aren't actually referenced anywhere explicitly though, because they're only ever put into the property bag?

I think the issue is that we have existing `v4` types that have a `v3` thing inside of their Property Bag? But if v3 and v4 always have the same structure then that maybe doesn't matter either?

I am not following why this reverses our current direction. Wouldn't we implement conversion for this new type exactly the same way we implemented it for all the other types? It's just it wouldn't be referred to anywhere as a property? For example why couldn't we continue having the conversion code to convert to the `v4.Address` on the `v3.Address`? (and same for `v5` from `v4`)?